test: add tests for the `schedule` host fn

ddd-mtl opened the pull request:
https://github.com/holochain/holochain/issues/5139

Added simple tests to check if `schedule` host fn is behaving as expected.

3 tests are ignored as they currently don't pass:
 - When a persisted function returns `None` it should be unscheduled.
 - When a persisted function fails for whatever reason it should be unscheduled.
 - When a persisted function provides an invalid crontab, it should be unscheduled.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Documentation**
  * Clarified scheduling behavior (ephemeral vs. persisted), trigger inputs, error handling, invalid crontab handling, and improved readability; removed outdated guidance; updated changelog to note new scheduling tests.
* **Tests**
  * Added a comprehensive scheduling test suite covering ephemeral and persisted flows, error and edge cases, helpers to observe scheduling state and signals, and conditional inclusion behind a feature flag.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ddd-mtl commented on crates/holochain/tests/tests/schedule.rs:
```diff
@@ -0,0 +1,336 @@
+        "coordinator",
+        "ephemeral",
+        move |api, input| {
+            let _ = api.emit_signal(AppSignal::new(ExternIO::encode(input.clone()).unwrap()));
```
function emits a signal of its input as of way of signaling it has been called.

ddd-mtl commented on crates/holochain/tests/tests/schedule.rs:
```diff
@@ -0,0 +1,336 @@
+            match ms {
+                0 => Ok(None),
+                _ => Ok(Some(Schedule::Ephemeral(std::time::Duration::from_millis(
+                    ms - 1,
```
Decrease duration by 1ms after each call and stop at zero.

veeso commented on crates/holochain/tests/tests/schedule.rs:
```diff
@@ -0,0 +1,286 @@
+        .await;
+
+    // Scheduled function is first called with None input
+    assert_eq!(None, wait_for_signal(&mut app_signal).await.unwrap());
```
```suggestion
    assert!(wait_for_signal(&mut app_signal).await.is_none());
```

ThetaSinner replied:
That is more compact and I was wondering about it above but actually it's nice to see what's inside the `Some` if this fails

veeso commented on crates/holochain/tests/tests/schedule.rs:
```diff
@@ -0,0 +1,286 @@
+        .await;
+
+    // Scheduled function is first called with None input.
+    assert_eq!(None, wait_for_signal(&mut app_signal).await.unwrap());
```
```suggestion
    assert!(wait_for_signal(&mut app_signal).await.is_none());
```

ThetaSinner commented on crates/holochain/tests/tests/mod.rs:
```diff
@@ -21,6 +21,8 @@ mod new_lair;
 mod publish;
 mod regression;
+#[cfg(feature = "unstable-functions")]
+mod schedule;
```
This is a good place to add tests like these. There are a lot of sweettests mixed into the codebase where ideally only unit tests would be and it makes it slow to run just the unit tests on Holochain. I'd really like to move all our integration tests out here.

ddd-mtl replied:
Does that work for you if I move the pre-existing schedule tests in that file in a subsequent PR?

ThetaSinner commented on crates/holochain/tests/tests/schedule.rs:
```diff
@@ -0,0 +1,286 @@
+    let zome = create_schedule_zome(|api, input| {
+        let _ = api.emit_signal(AppSignal::new(ExternIO::encode(input.clone()).unwrap()));
+        let ms: u64 = match input {
+            Some(Schedule::Ephemeral(duration)) => duration.as_millis() as u64,
+            None => 3,
```
```suggestion
            None => 3,
            Some(Schedule::Ephemeral(duration)) => duration.as_millis() as u64,
```

Possibly easier to follow? The `None` is expected first, then we take the input value on subsequent schedules calls.

ThetaSinner reviewed (commented):
The testing approach looks sound. What I'm unclear about is why all the tests pass. Are the tests being written to pass what is currently happening or is there missing test coverage for the problem cases?

Increasing test coverage was the point of this so if the problem cases covered then that's okay. They could be added as ignored but they could also be added later.

If the tests are testing the current state of things and the assertions aren't matching the desired end state then I'd prefer assertions that fail and to just mark the tests as `#[ignore]` for now until we fix the problems.

ThetaSinner commented:
> When a persisted function returns None it should be unscheduled.

Yes

> An ephemeral duration of 0 is deemed acceptable and the function stays scheduled.

I think that is a little odd but probably acceptable if documented.

> A persisted function that fails for whatever reason will still stay scheduled.

That's a tricky one because it could be that the error is systemic and will just keep happening so we'd be better off unscheduling. That's a more complicated change if we need to modify the required return values and behavior of scheduled functions. I think this would be worth some investigation for what would need to change and then discussion?

> When a persisted function changes to provide an invalid crontab, it will still stay scheduled.

Again, hard to tell whether the problem is systemic or temporary. I think demanding valid crontab being returned is reasonable and the schedule should be cancelled.

ddd-mtl commented:
@ThetaSinner Changed tests to match expectations and set to `ignore` when failing. Changed doc accordingly.

Added ephemeral_error test and persisted_expired test to cover more cases.

matthme reviewed (commented):
Changelog in `crates/holochain/CHANGELOG.md` is missing. It's "only" adding tests but I think it's still worth mentioning in the changelog.
